core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -117,6 +156,32 @@ public void setDisposer(@NonNull Disposer disposer) {
      * Must be safely serializable, so it receives runtime context comparable to that of the original setup.
      */
     public static abstract class Disposer implements Serializable {
+
+        /**
+         * Creates a new end-of-wrapped-block callback.
+         *
+         * @deprecated Use the constructor that takes a wrapper instance instead.
```

Comment:
Should not be necessary. `Context` can be told whether its associated wrapper requests a workspace; then `setDisposer` can pass this bit of information along to the `Disposer`.

Reply:
That is possible for non-pipeline cases only. When we're dealing with `AbstractBuild`, then the final overrides of `setUp()` and `preCheckout()` offer a place where I can tell the `Context` whether or not the wrapper needs a workspace context. Then I can just use a private `Boolean` field on both `Context` and `Disposer` to pass the flag across.

But for pipeline use, all methods involved are virtual and non-final, so there does not seem to be any place in code where I can guarantee interaction between the created `Context` and the `SimpleBuildWrapper` instance. The equivalent place in `CoreWrapperStep` would be the `doStart()` method. But then I would have to add public API to `Context` to flag the requirement there, which feels worse than simply allowing the `Dispatcher` to be told at construction time.

Reply:
> add public API to `Context` to flag the requirement there

Reply:
Are there really no use cases supported other than `AbstractBuild` or pipeline? If not, then I suppose it's fine as long as `workflow-basic-steps` sets things up correctly.

Reply:
> Are there really no use cases supported other than `AbstractBuild` or pipeline?

No, that is it.

Reply:
OK, I think I have something that works. Note that I think this does mean that when an old version of `workflow-basic-steps` is used with new core + a plugin with a wrapper marked as no-workspace-required, then any `Disposer` for that wrapper _will_ require that context and potentially fail (because the `Context.setWorkspaceRequirement()` won't have been called).